src/base/rss/feed_serializer.cpp

```diff
@@ -0,0 +1,125 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "feed_serializer.h"
+
+#include <QFile>
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QJsonObject>
+
+#include "base/logger.h"
+#include "base/path.h"
+#include "base/utils/io.h"
+#include "rss_article.h"
+
+void RSS::Private::FeedSerializer::load(const Path &dataFileName, const QString &url)
+{
+    QFile file {dataFileName.data()};
+
+    if (!file.exists())
+    {
+        emit loadingFinished({});
+    }
+    else if (file.open(QFile::ReadOnly))
+    {
+        emit loadingFinished(loadArticles(file.readAll(), url));
+        file.close();
+    }
+    else
+    {
+        LogMsg(tr("Couldn't read RSS Session data from %1. Error: %2")
+               .arg(dataFileName.toString(), file.errorString())
+               , Log::WARNING);
+    }
+}
+
+void RSS::Private::FeedSerializer::store(const Path &dataFileName, const QVector<QVariantHash> &articlesData)
+{
+    QJsonArray arr;
+    for (const QVariantHash &data : articlesData)
+    {
+        auto jsonObj = QJsonObject::fromVariantHash(data);
+        // JSON object doesn't support DateTime so we need to convert it
+        jsonObj[Article::KeyDate] = data[Article::KeyDate].toDateTime().toString(Qt::RFC2822Date);
+
+        arr << jsonObj;
+    }
+
+    const nonstd::expected<void, QString> result = Utils::IO::saveToFile(dataFileName, QJsonDocument(arr).toJson());
+    if (!result)
+    {
+       LogMsg(tr("Failed to save RSS feed in '%1', Reason: %2").arg(dataFileName.toString(), result.error())
+              , Log::WARNING);
+    }
+}
+
+QVector<QVariantHash> RSS::Private::FeedSerializer::loadArticles(const QByteArray &data, const QString &url)
+{
+    QJsonParseError jsonError;
+    const QJsonDocument jsonDoc = QJsonDocument::fromJson(data, &jsonError);
+    if (jsonError.error != QJsonParseError::NoError)
+    {
+        LogMsg(tr("Couldn't parse RSS Session data. Error: %1").arg(jsonError.errorString())
+               , Log::WARNING);
+        return {};
+    }
+
+    if (!jsonDoc.isArray())
+    {
+        LogMsg(tr("Couldn't load RSS Session data. Invalid data format."), Log::WARNING);
+        return {};
+    }
+
+    QVector<QVariantHash> result;
+    const QJsonArray jsonArr = jsonDoc.array();
+    int i = -1;
+    for (const QJsonValue &jsonVal : jsonArr)
```

Comment:
I would use index instead:
```c++
for (int i = 0; i < jsonArr.size(); ++i)
{
    const QJsonValueRef jsonVal = jsonArr[i];
    // ...
```

Reply:
Is there any sense to use QJsonValueRef if you don't want to modify it? Maybe QJsonValue?

Reply:
>Is there any sense to use QJsonValueRef if you don't want to modify it? 

As usual, it prevents a copy.

Reply:
> > Is there any sense to use QJsonValueRef if you don't want to modify it?
> 
> As usual, it prevents a copy.

Using `QJsonValue` doesn't actually copy anything due to implicitly sharing. `QJsonValueRef` is intended to be used in mutable context (instead of `QJsonValue &`, since the latter can't be used for some reason).
Anyway `QJsonArray::operator[](int i) const` returns `QJsonValue`.

Reply:
OK I don't really insist it.
I would like to note that implicitly sharing is still slower than taking a reference `&`.